### PR TITLE
book: explanation that String -> &str coercion doesn't happen for &str traits

### PR DESCRIPTION
**Pull request by thombles**
A few of us [over on the forum](https://users.rust-lang.org/t/string-type-coercion-in-rust/1439) have been tripped up by this distinction, which I don't think is mentioned. It's kind of logical if you read the "Deref coercions" page and squint a bit but I think it would be nice to explain it directly. Here's one way we could clarify it.

### CODE REVIEW

File: src/doc/trpl/strings.md [hunk 1]
```diff
@@ -49,6 +49,18 @@ fn main() {
 }
 ```
 
+This coercion does not happen for functions that accept one of `&str`’s traits
+instead of `&str`. For example, [`TcpStream::connect`][connect] has a parameter
+of type `ToSocketAddrs`. A `&str` is okay but a `String` must be explicitly
+converted using `&*`.
+
+```rust
```
Comment (steveklabnik) [inline src/doc/trpl/strings.md]: this should probably be tagged `no_run` as well, because it won't work, since there's no server, yeah?

**Comment by steveklabnik** [comment]
Thanks so much! r=me after the nit

**Comment by thombles** [comment]
I've added `no_run` and a use statement so it will build. Not a bad thing to have anyway. `check-docs` is all good now.

**Comment by steveklabnik** [comment]
Thanks so much!

**Pull request merged**